### ISSUE TITLE
RLP Problem with small lists

### ISSUE THREAD

**Issue by ligi** [web3j/web3j#124]
I am just writing the RLP encoder in Kotlin and while doing this stumbling around one corner case that web3j also fails - simple example:
  `RlpEncoder.encode(new RlpList(RlpString.create("a")).size`

returns 1 - but should be 2

Fix can be something like this:

From:
```
 private static byte[] encode(byte[] bytesValue, int offset) {
        if (bytesValue.length == 1
                && bytesValue[0] >= (byte) 0x00
                && bytesValue[0] <= (byte) 0x7f) {
            return bytesValue;
```

to:
```
 private static byte[] encode(byte[] bytesValue, int offset) {
        if (bytesValue.length == 1
                && offset==STRING_OFFSET
                && bytesValue[0] >= (byte) 0x00
                && bytesValue[0] <= (byte) 0x7f) {
            return bytesValue;
```

----

I noticed this problem when implementing this test:
```
  "multilist": {
    "in": [ "zw", [ 4 ], 1 ],
    "out": "c6827a77c10401"
  },
```

from: https://github.com/ethereum/tests/blob/develop/RLPTests/rlptest.json

FYI: here the Kotlin implementation: https://github.com/walleth/kethereum/commit/20e2265ba6ec2805a3cae148f610b607ce006b14?diff=unified

**Comment by ligi** [comment]
You are very welcome!

There are a lot of reasons for a Kotlin Etherum implementation - some from the top of my head:
  
 * Kotlin does not only target JVM - Kotlin can also target JavaScript and (soon) native - this could get very interesting in the future 
 * I love small libraries - web3j is already kinda big and doing a lot of things I don't need - I split kethereum into many modules from the beginning
 * The code is much more concise (e.g. just compare the RLP implementation in Java vs. Kotlin) and so e.g. easier to review (also in combination with it being modularized)
 * I kinda don't want to write java anymore (saying this after writing java for years with pain reaching back to J2ME) - If I would use web3j I see myself having to write java again (e.g. to fix issues I stumble uppon)
 * I learn a lot about Ethereum under the hood by writing this code which I would not if I just see it through an API
 * it's fun
 * ..

**Comment by conor10** [comment]
Fair enough - thanks for providing your thoughts on this @ligi. I'll be doing some work with web3j and Kotlin shortly. 

Agree - writing OSS is fun :)

**Comment by conor10** [comment]
Fixed.

**Comment by conor10** [comment]
Regarding the above - are there specific parts of web3j that would be useful to you if more modular? I'm looking at modularising web3j and providing some smaller artefacts with releases.

**Comment by rmeissner** [comment]
Even so it is a little late I do have some feedback on that @conor10  😄 

We were actually also looking into using web3j but even now that it is modularised it feels too big. We would only need the classes for accounts and signing, but there is no way as long as there are not a lot of small modules (as we don'T want to pull JavaPoet or Mockito). 

Also when developing an app for Android bouncycastle might become problematic (see https://rtyley.github.io/spongycastle/) . 

Because of this we decided to implement our own library.

**Comment by conor10** [comment]
Thanks for the feedback @rmeissner - it's much appreciated. I'll break it up further for the 3.0 release. 

The Android fork has always used Spongcastle - I've updated the main page to state this.

**Comment by ligi** [comment]
Sorry - completely overlooked this request - just saw it via the recent comment of @rmeissner and the closing of #146 . First of all thanks for the fix(es)!
regarding spongycastle have a look at:
* https://github.com/bitcoinj/bitcoinj/pull/1440
* https://android.googlesource.com/platform/libcore/+/0ac85ead96f1ba7d35f3acadd154de4ef0a8fd87%5E%21/

TLDR: with the right minSDK you do not need to use Spongycastle anymore

And I am all for more modules - looking forward to 3.0!

**Comment by conor10** [comment]
I've broken up web3j into the core modules I had originally been thinking of.

* core
* codegen
* console
* geth
* parity
* infura
* integration-test

Question - would you find it useful to have rlp, abi (& potentially crypto) as published web3j artifacts?

**Comment by ligi** [comment]
Sounds great - and IMHO yes - the more modules the better ;-)
